infra/src/main/kotlin/fi/fta/geoviite/infra/geometry/VerticalGeometryListing.kt

```diff
@@ -0,0 +1,158 @@
+    val stationPoint = circCurveStationPoint(segment)
+    val stationPointCoordinates = alignment.getCoordinateAt(alignment.stationValueNormalized(stationPoint.x))
+        ?.let { coordinateTransform?.transform(it) }
+    val startCoordinates = alignment.getCoordinateAt(alignment.stationValueNormalized(segment.start.x))
+        ?.let { coordinateTransform?.transform(it) }
+    val endCoordinates = alignment.getCoordinateAt(alignment.stationValueNormalized(segment.end.x))
+        ?.let { coordinateTransform?.transform(it) }
+
+    return VerticalGeometryListing(
+        id = StringId("${alignment.id}_${segment.start.x}"),
+        planId = planId,
+        planSource = planSource,
+        fileName = planFileName,
+        alignmentId = alignment.id,
+        alignmentName = alignment.name,
+        null,
+        start = CurvedSectionEndpoint(
+            address = startCoordinates?.let { geocodingContext?.getAddress(startCoordinates)?.first },
+            height = segment.start.y,
+            angle = angleFractionBetweenPoints(stationPoint, segment.start),
+            station = segment.start.x
+        ),
+        end = CurvedSectionEndpoint(
+            address = endCoordinates?.let { geocodingContext?.getAddress(endCoordinates)?.first },
+            height = segment.end.y,
+            angle = angleFractionBetweenPoints(stationPoint, segment.end),
+            station = segment.end.x
+        ),
+        point = IntersectionPoint(
+            address = stationPointCoordinates?.let { geocodingContext?.getAddress(stationPointCoordinates)?.first },
+            height = stationPoint.y,
+            station = stationPoint.x
+        ),
+        radius = segment.radius,
+        tangent = lineLength(segment.start, stationPoint),
+        linearSectionBackward = previousLinearSection(segment, curvedSegments, linearSegments),
+        linearSectionForward = nextLinearSection(segment, curvedSegments, linearSegments),
+    )
+}
+
+fun previousLinearSection(
+    currentSegment: CurvedProfileSegment,
+    curvedSegments: List<CurvedProfileSegment>,
+    linearSegments: List<LinearProfileSegment>,
+): LinearSection {
+    val previousCurvedSegment = curvedSegments.findLast { it.start.x < currentSegment.start.x }
+    val previousLinearSegment =
+        linearSegments.findLast { it.start.x < currentSegment.start.x && (previousCurvedSegment == null || it.start.x > previousCurvedSegment.start.x) }
+    return LinearSection(
+        linearSegmentLength = previousLinearSegment?.let { previousLinearSegment.end.x - previousLinearSegment.start.x },
+        stationValueDistance = previousLinearSegment?.let {
+            val currentStationPoint = circCurveStationPoint(currentSegment)
+            val previousStationPoint = previousCurvedSegment?.let { prev -> circCurveStationPoint(prev) }
+                ?: previousLinearSegment.start
+            currentStationPoint.x - previousStationPoint.x
+        }
+    )
+}
```

Comment:
Kaltevuusjakso taaksepäin. Näiden speksit kannattaa katsoa mirosta: https://miro.com/app/board/uXjVP4Z0Mn8=/

Kaltevuusjakso kuvaa siis kaltevuudeltaan muuttumatonta pätkää kahden pyöristyksen välillä. Otin näissä itse pari taiteellista speksin ulkopuolista vapautta:
- Jos on kaksi pyöristystä peräkkäin, kaltevuusjaksoa ei ole (arvot nulleja)
- Jos kaltevuusjakson toinen pää ei ole kiinni pyöristyksessä (esim. jos alignment alkaa tai loppuu kaltevuusjaksoon), kaltevuusjakson pituus lasketaan kaltevuusjakson ka. pään lopusta. Pyöristysten välissä se on speksattu laskettavaksi (ja lasketaan) pyöristysten taitepisteiden välisestä etäisyydestä
- Tässä tutkitaan vain yksittäistä kaltevuusjaksoa. Useassa peräkkäisessä kaltevuusjaksossa ei sinällään pitäisi olla järkeä, joten niihin ei ole varauduttu (miksi olisi monta putkeen kun ne on samaa kaltevuutta muutenkin? Tai jos ei ole, niin niiden välistä puuttuu pyöristys.) Datasta en tosin ole tutkinut sattuisiko olemaan monta putkeen

Reply:
Datassa niitä ei pitäisi olla koska im:n esitysmuodosta sellaista ei voi syntyä ellei pvi-pisteitä ole käytetty keskellä profiilia. Muualla laskennassa onvarauduttu siihenkin, joten yhtenäisyyden puolesta voisi periaatteessa olla myös tässä, mutta käytännössä sellaista ei pitäisi kyllä olla. Sitten kun näitä katsotaan paikannuspohjan raiteesta, tuo lienee mahdollista jos suunnitelma vaihtuu.